### PR TITLE
librbd: possible deadlock if cluster connection closed after image

### PR DESCRIPTION
**Pull request by dillaman**
Fixes: http://tracker.ceph.com/issues/17254
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

### CODE REVIEW

**Comment by xiaoxichen** [comment]
looks good, hoping this  in 10.2.3

**Comment by xiaoxichen** [comment]
retest this please

**Comment by dillaman** [comment]
@xiaoxichen If I created a 10.2.2 dev packages with this fix, would you be able to verify that it solves the problem?

**Comment by xiaoxichen** [comment]
@dillaman of course, YES!

**Comment by dillaman** [comment]
@xiaoxichen Trusty v10.2.2 + this patch is available under the wip-17254-jewel branch. CentOS 7 builder failed due to environment issues.

**Comment by xiaoxichen** [comment]
@dillaman ,thx, we are all in ubuntu trusty:)

**Comment by xiaoxichen** [comment]
@dillaman ,deployed the patched library to the cluster. Lets bake and see.

File: src/librbd/ImageWatcher.cc [hunk 1]
```diff
@@ -55,13 +55,19 @@ struct C_UnwatchAndFlush : public Context {
...
+      return;
     }
+
+    // ensure our reference to the RadosClient is released prior
+    // to completing the callback to avoid racing an explicit
+    // librados shutdown
+    Context *ctx = on_finish;
+    delete this;
+
+    ctx->complete(r);
```
Comment (trociny) [inline src/librbd/ImageWatcher.cc]: @dillaman Shouldn't it complete with ret_val?

**Comment by xiaoxichen** [comment]
@dillaman , it goes well in our enviroment for more than 12 hours, looking good except it should complete with ret_val

File: src/librbd/ImageWatcher.cc [hunk 2]
```diff
@@ -55,13 +55,19 @@ struct C_UnwatchAndFlush : public Context {
...
+      return;
     }
+
+    // ensure our reference to the RadosClient is released prior
+    // to completing the callback to avoid racing an explicit
+    // librados shutdown
+    Context *ctx = on_finish;
+    delete this;
+
+    ctx->complete(ret_val);
```
Comment (trociny) [inline src/librbd/ImageWatcher.cc]: @dillaman Ah, missed this: we can't use ret_val here. Need to copy it before delete.

**Pull request merged**